diff -ruN ../originalVersion/OpenFOAM-6/src/Pstream/dummy/UPstream.C ../finalVersion/OpenFOAM-6/src/Pstream/dummy/UPstream.C
--- ../originalVersion/OpenFOAM-6/src/Pstream/dummy/UPstream.C	2019-08-22 17:01:27.398307000 +0100
+++ ../finalVersion/OpenFOAM-6/src/Pstream/dummy/UPstream.C	2019-08-22 17:15:32.638170400 +0100
@@ -32,7 +32,7 @@
 {}
 
 
-bool Foam::UPstream::init(int& argc, char**& argv, const bool needsThread)
+bool Foam::UPstream::init(int& argc, char**& argv, const bool needsThread, bool coupled)
 {
     FatalErrorInFunction
         << "Trying to use the dummy Pstream library." << nl
diff -ruN ../originalVersion/OpenFOAM-6/src/Pstream/mpi/Make/options ../finalVersion/OpenFOAM-6/src/Pstream/mpi/Make/options
--- ../originalVersion/OpenFOAM-6/src/Pstream/mpi/Make/options	2019-08-22 17:01:27.409309200 +0100
+++ ../finalVersion/OpenFOAM-6/src/Pstream/mpi/Make/options	2019-08-22 17:15:32.647194600 +0100
@@ -1,5 +1,6 @@
 sinclude $(GENERAL_RULES)/mplib$(WM_MPLIB)
 sinclude $(RULES)/mplib$(WM_MPLIB)
+sinclude $(GENERAL_RULES)/MUI
 
-EXE_INC  = $(PFLAGS) $(PINC)
+EXE_INC  = $(PFLAGS) $(PINC) $(MUI_INC)
 LIB_LIBS = $(PLIBS)
diff -ruN ../originalVersion/OpenFOAM-6/src/Pstream/mpi/PstreamGlobals.C ../finalVersion/OpenFOAM-6/src/Pstream/mpi/PstreamGlobals.C
--- ../originalVersion/OpenFOAM-6/src/Pstream/mpi/PstreamGlobals.C	2019-08-22 17:01:27.413821200 +0100
+++ ../finalVersion/OpenFOAM-6/src/Pstream/mpi/PstreamGlobals.C	2019-08-22 17:15:32.651707800 +0100
@@ -34,6 +34,8 @@
 
 MPI_Comm PstreamGlobals::MPI_COMM_FOAM;
 
+MPI_Comm PstreamGlobals::commWorld_;
+
 // Outstanding non-blocking operations.
 //! \cond fileScope
 DynamicList<MPI_Request> PstreamGlobals::outstandingRequests_;
diff -ruN ../originalVersion/OpenFOAM-6/src/Pstream/mpi/PstreamGlobals.H ../finalVersion/OpenFOAM-6/src/Pstream/mpi/PstreamGlobals.H
--- ../originalVersion/OpenFOAM-6/src/Pstream/mpi/PstreamGlobals.H	2019-08-22 17:01:27.418360500 +0100
+++ ../finalVersion/OpenFOAM-6/src/Pstream/mpi/PstreamGlobals.H	2019-08-22 17:15:32.661733400 +0100
@@ -59,6 +59,9 @@
 
     extern DynamicList<int> freedTags_;
 
+    //Default MPI communicator (either allocated by MUI if in use or becomes MPI_COMM_WORLD)
+    extern MPI_Comm commWorld_;
+
     // Current communicators. First element will be MPI_COMM_FOAM
     extern DynamicList<MPI_Comm> MPICommunicators_;
 
diff -ruN ../originalVersion/OpenFOAM-6/src/Pstream/mpi/UPstream.C ../finalVersion/OpenFOAM-6/src/Pstream/mpi/UPstream.C
--- ../originalVersion/OpenFOAM-6/src/Pstream/mpi/UPstream.C	2019-08-22 17:01:27.432372000 +0100
+++ ../finalVersion/OpenFOAM-6/src/Pstream/mpi/UPstream.C	2019-08-22 17:15:32.665242600 +0100
@@ -40,6 +40,12 @@
 #include <cstdlib>
 #include <csignal>
 
+// NOTE:
+// MUI header included if the switch -DUSE_MUI included during compilation.
+#ifdef USE_MUI
+    #include "mui.h"
+#endif
+
 #if defined(WM_SP)
     #define MPI_SCALAR MPI_FLOAT
 #elif defined(WM_DP)
@@ -65,38 +71,59 @@
 }
 
 
-bool Foam::UPstream::init(int& argc, char**& argv, const bool needsThread)
+bool Foam::UPstream::init(int& argc, char**& argv, const bool needsThread, bool coupled)
 {
-    // MPI_Init(&argc, &argv);
-    int provided_thread_support;
-    MPI_Init_thread
-    (
-        &argc,
-        &argv,
-        (
-            needsThread
-          ? MPI_THREAD_MULTIPLE
-          : MPI_THREAD_SINGLE
-        ),
-        &provided_thread_support
-    );
-
-    // int numprocs;
-    // MPI_Comm_size(MPI_COMM_WORLD, &numprocs);
-    // int myRank;
-    // MPI_Comm_rank(MPI_COMM_WORLD, &myRank);
-
-    int myGlobalRank;
-    MPI_Comm_rank(MPI_COMM_WORLD, &myGlobalRank);
-    MPI_Comm_split
-    (
-        MPI_COMM_WORLD,
-        1,
-        myGlobalRank,
-        &PstreamGlobals::MPI_COMM_FOAM
-    );
+	int provided_thread_support;
+
+	if(coupled)
+	{
+		#ifdef USE_MUI
+			//Use world returned by MUI, based on MPI MPMD model, calls MPI_Init_thread inside MUI function
+			PstreamGlobals::commWorld_ = mui::mpi_split_by_app(argc, argv, (needsThread ? MPI_THREAD_MULTIPLE : MPI_THREAD_SINGLE), &provided_thread_support);
+		#else
+			PstreamGlobals::commWorld_ = MPI_COMM_WORLD;
+
+			MPI_Init_thread
+			(
+				&argc,
+				&argv,
+				(
+					needsThread
+				  ? MPI_THREAD_MULTIPLE
+				  : MPI_THREAD_SINGLE
+				),
+				&provided_thread_support
+			);
+		#endif
+	}
+	else
+	{
+		PstreamGlobals::commWorld_ = MPI_COMM_WORLD;
+
+		MPI_Init_thread
+		(
+			&argc,
+			&argv,
+			(
+				needsThread
+			  ? MPI_THREAD_MULTIPLE
+			  : MPI_THREAD_SINGLE
+			),
+			&provided_thread_support
+		);
+	}
+
+	int myGlobalRank;
+	MPI_Comm_rank(PstreamGlobals::commWorld_, &myGlobalRank);
+	MPI_Comm_split
+	(
+		PstreamGlobals::commWorld_,
+		1,
+		myGlobalRank,
+		&PstreamGlobals::MPI_COMM_FOAM
+	);
 
-    int numprocs;
+	int numprocs;
     MPI_Comm_size(PstreamGlobals::MPI_COMM_FOAM, &numprocs);
     int myRank;
     MPI_Comm_rank(PstreamGlobals::MPI_COMM_FOAM, &myRank);
diff -ruN ../originalVersion/OpenFOAM-6/src/OpenFOAM/db/IOstreams/Pstreams/UPstream.H ../finalVersion/OpenFOAM-6/src/OpenFOAM/db/IOstreams/Pstreams/UPstream.H
--- ../originalVersion/OpenFOAM-6/src/OpenFOAM/db/IOstreams/Pstreams/UPstream.H	2019-08-22 17:01:19.688652300 +0100
+++ ../finalVersion/OpenFOAM-6/src/OpenFOAM/db/IOstreams/Pstreams/UPstream.H	2019-08-22 17:15:32.555950600 +0100
@@ -366,7 +366,7 @@
 
         //- Initialisation function called from main
         //  Spawns slave processes and initialises inter-communication
-        static bool init(int& argc, char**& argv, const bool needsThread);
+        static bool init(int& argc, char**& argv, const bool needsThread, bool coupled=false);
 
         // Non-blocking comms
 